components/payment.js

```diff
@@ -0,0 +1,211 @@
+          .then(resolve)
+          .catch(reject)
+        invoice.waitUntilPaid(id)
+          .then(resolve)
+          .catch(reject)
+      })
+    } catch (err) {
+      console.error('WebLN payment failed:', err)
+      throw err
+    }
+  }, [provider, invoice])
+
+  return waitForWebLnPayment
+}
+
+const useQrPayment = () => {
+  const invoice = useInvoice()
+  const showModal = useShowModal()
+
+  const waitForQrPayment = useCallback(async (inv) => {
+    return await new Promise((resolve, reject) => {
+      let paid
+      const cancelAndReject = async (onClose) => {
+        if (paid) return
+        await invoice.cancel(inv)
+        reject(new InvoiceCanceledError(inv.hash))
+      }
+      showModal(onClose =>
+        <Invoice
+          invoice={inv}
+          modal
+          successVerb='received'
+          webLn={false}
+          onPayment={() => { paid = true; onClose(); resolve() }}
+          poll
+        />,
+      { keepOpen: true, onClose: cancelAndReject })
+    })
+  }, [invoice])
+
+  return waitForQrPayment
+}
+
+export const PaymentProvider = ({ children }) => {
+  const me = useMe()
+  const feeButton = useFeeButton()
+  const invoice = useInvoice()
+  const waitForWebLnPayment = useWebLnPayment()
+  const waitForQrPayment = useQrPayment()
+
+  const waitForPayment = useCallback(async (invoice) => {
+    try {
+      return await waitForWebLnPayment(invoice)
+    } catch (err) {
+      if (err instanceof InvoiceCanceledError) {
+        // bail since qr code payment will also fail if invoice was canceled
+        throw err
+      }
+      // ignore any other error and fallback to QR code
+    }
+    return await waitForQrPayment(invoice)
+  }, [waitForWebLnPayment, waitForQrPayment])
+
+  const request = useCallback(async (amount) => {
+    amount ??= feeButton?.total
+    const free = feeButton?.free
+
+    // if user has enough funds in their custodial wallet or action is free, never prompt for payment
+    // XXX this will probably not work as intended for deposits < balance
+    //   which means you can't always fund your custodial wallet with attached wallets ...
+    //   but should this even be the case?
+    const insufficientFunds = !me || me.privates.sats < amount
+    if (free || !insufficientFunds) return [{ hash: null, hmac: null }, null]
+
+    const inv = await invoice.create(amount)
+
+    await waitForPayment(inv)
+
+    const cancel = () => invoice.cancel(inv).catch(console.error)
+    return [inv, cancel]
+  }, [me, feeButton?.total, invoice, waitForPayment])
+
+  const cancel = useCallback(({ hash, hmac }) => {
+    if (hash && hmac) {
+      invoice.cancel({ hash, hmac }).catch(console.error)
+    }
+  }, [invoice])
+
+  const value = useMemo(() => ({ request, cancel }), [request, cancel])
+  return (
+    <PaymentContext.Provider value={value}>
+      {children}
+    </PaymentContext.Provider>
+  )
+}
+
+export const usePayment = () => {
+  return useContext(PaymentContext)
+}
```

Comment:
Mhh, I don't really need to wrap the functions in a context since I do nothing with it. I thought I might need to be context aware but even better that I actually don't. I always return the same two functions. Guess I can remove it and move all code from `PaymentProvider` to `usePayment`.